sros2/test/sros2/policy/test_permission.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2019 Canonical Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import textwrap
+
+from lxml import etree
+
+import pytest
+
+from sros2.policy import _expression, _permission
+
+
+def test_topic_publish_allow_from_xml():
+    permission_xml = etree.fromstring(textwrap.dedent("""\
+        <topics publish="ALLOW">
+            <topic>test_topic</topic>
+        </topics>
+    """))
+
+    permission = _permission.Permission(permission_xml)
+    _assert_topic_publish(permission)
+    _assert_topic_expressions(permission)
+
+
+def test_topic_subscribe_allow_from_xml():
+    permission_xml = etree.fromstring(textwrap.dedent("""\
+        <topics subscribe="ALLOW">
+            <topic>test_topic</topic>
+        </topics>
+    """))
+
+    permission = _permission.Permission(permission_xml)
+    _assert_topic_subscribe(permission)
+    _assert_topic_expressions(permission)
+
+
+def test_service_reply_allow_from_xml():
+    permission_xml = etree.fromstring(textwrap.dedent("""\
+        <services reply="ALLOW">
+            <service>test_service</service>
+        </services>
+    """))
+
+    permission = _permission.Permission(permission_xml)
+    _assert_service_reply(permission)
+    _assert_service_expressions(permission)
+
+
+def test_topic_publish_allow_from_fields():
+    _assert_topic_publish(_permission.Permission.from_fields(
+        _permission.PermissionType.TOPIC, _permission.PermissionRuleType.PUBLISH,
+        _permission.PermissionRuleQualifier.ALLOW))
+
+
+def test_topic_subscribe_allow_from_fields():
+    _assert_topic_subscribe(_permission.Permission.from_fields(
+        _permission.PermissionType.TOPIC, _permission.PermissionRuleType.SUBSCRIBE,
+        _permission.PermissionRuleQualifier.ALLOW))
+
+
+def test_service_reply_allow_from_fields():
+    _assert_service_reply(_permission.Permission.from_fields(
+        _permission.PermissionType.SERVICE, _permission.PermissionRuleType.REPLY,
+        _permission.PermissionRuleQualifier.ALLOW))
+
+
+def test_unsupported_permission():
+    permission_xml = etree.fromstring(textwrap.dedent("""\
+        <topics publish="ALLOW" subscribe="ALLOW">
```

Comment:
Why is this unsupported?
These attributes are not purely exclusive.

https://github.com/ros2/sros2/blob/1c274b41cde9ab30220b85b1b82d164dae0d23a1/sros2/sros2/policy/schemas/policy.xsd#L41-L42

Although, for sros, at least one should be set.
But I'm not sure that is expressible in XSD 1.0.
I could be wrong, but the current xsd doesn't seem to enforce this.

Reply:
Mostly because the existing code didn't write XML that had them both, but I realize now that we definitely need to be able to at least _parse_ XML that does, even if we don't write them. I'll fix it.